[RFC]PyTorchTVM (#25)
apache/tvm-rfcs: rfcs/0025-add-pytorch-tvm.md

- Feature Name: PyTorchTVM
- Start Date: 2021-08-24
- RFC PR: [apache/tvm-rfcs#0025](https://github.com/apache/tvm-rfcs/pull/25)
- GitHub Issue: TODO

# Summary
[summary]: #summary

> This RFC add a `PyTorchTVM` module to support: compile TorchScript to TVM and use accelerated module in PyTorch.

Reviewer: Just wanted to make sure I understand what it exactly means, let me know if the sentence below is consistent with the proposal. Thanks!

```suggestion
This RFC add a `PyTorchTVM` module to support: offload subgraphs of TorchScript to TVM, and then embed those TVM-accelerated subgraphs back to TorchScript for runtime execution.
```

> To increase the TVM accessibility for PyTorch users, we propose `PyTorchTVM` module to support the following workflow:

Reviewer: ```suggestion
To help boost model performance and enhance TVM adoption for machine learning practitioners who often use PyTorch, `PyTorchTVM` is proposed for seamless integration for TVM in TorchScript, and its workflow is demonstrated as follows:
```

> 1. convert a torchscript module to tvm graph

Reviewer: To clarify, is it an entire module (full graph) or a subgraph?

```suggestion
1. Convert a TorchScript module to TVM graph (Relay)
```

Author: A torchscript module contains other torchscript module. So users can choose a module at any level to do the conversion.

Reviewer: what about

```suggestion
1. Convert a TorchScript module (or submodule) to TVM graph (Relay)
```

> 2. build and tune tvm graph

Reviewer: ```suggestion
2. Optimize and compile the TVM graph with auto-tuning
```

> 3. export well-tuned tvm graph as a pytorch op

Reviewer: ```suggestion
3. Export and embed the optimized TVM module as a PyTorch custom op
```

> 4. torch jit trace the tvm pytorch op with other pytorch modules, then save/load/serve as normal pytorch model

Reviewer: ```suggestion
4. The embedded custom op works smoothly with TorchScript (the `torch.jit.trace` API), without tangible difference with normal PyTorch models, i.e. it can be saved to disk, loaded back and served online with no change in the overall workflow
```

Author: Actually works with both the `torch.jit.trace` and `torch.jit.script` API

Reviewer: Okay this is important. We should state really clearly at the beginning saying it works with both APIs

Reviewer: ```suggestion
4. The embedded custom op works smoothly with TorchScript (both `torch.jit.trace` and `torch.jit.script`), without tangible difference with normal PyTorch models, i.e. it can be saved to disk, loaded back and served online with no change in the overall workflow
```

# Motivation
[motivation]: #motivation

> PyTorch framework is increasingly being adopted for research and production. At the same time, PyTorch lacks an effective inference acceleration toolchain, which is the main concern in the industry. Existing acceleration includes:

Reviewer: PyTorch enjoys increasing popularity among machine learning research community as well as in industrial production environment. However, it is still a missing piece as a generic, comprehensive and effective toolchain to accelerate real-world models and workloads in PyTorch, which raises primary concern in performance-critical production environments.

* PyTorch → ONNX → TensorRT/TVM
* PyTorch → torchscript → TensorRT/TVM

From our perspective, there are some limitations for both ONNX and TensorRT:

* Onnx cannot cover all models with dynamic control flow (e.g. for loop)

> * TensorRT can only accelerate some standard networks

Reviewer: Below are the two classic acceleration workflows as the status quo:
- PyTorch -> ONNX -> TensorRT/TVM
- PyTorch -> TorchScript -> TensorRT/TVM

However, both workflows introduce workflow introduce one level of indirection, which means flaws of either levels are inherited in the pipeline. For example:
- ONNX offers no support for models with dynamic control flow, so the first workflow is unable to support dynamic models
- The coverage of TensorRT is often limited to a range of standard neural networks, so both of the workflows, if offloaded to TensorRT, are hard to be effective on real-world irregular models.

Furthermore, both of the existing workflows don't provide any benefit of an interface that is practical enough for researchers to widely adopt and reuse. For example, it requires deep knowledge of TVM runtime modules to load the exported binary artifacts back to python and use it together with PyTorch.

Author: 1.  is `workflows introduce workflow introduce` a typo?
2. `dynamic models` is a little confusing, maybe we just use `models with dynamic control flow`
3. `real-world irregular models`, actually most models other than `conv` model cannot be accelerated very well, like `bert` or `wide and deep` models, which are not irregular.

Reviewer: 1. Yep of course it's a typo. Just got it fixed. Thanks!
2. Good point! I didn't realize I invent a weird word.
3. Huh okay, I was thinking about irregular shapes (like tall/skinny matmuls), but I agree here is some ambiguity. Removed the term "irregular"

Author: Emm, recently I found that onnx supported dynamic control flow. https://pytorch.org/docs/stable/onnx.html#tracing-vs-scripting

so we may need to rewrite this part.

> So we hope to use TVM to accelerate PyTorch model inference.

Reviewer: We should discuss more about the benefit of PyTorchTVM here

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

> For example, we have an end-to-end resnet classification model, consisting of 3 parts:
> 
> 1. Image reader
> 2. Image transforms
> 3. Resnet model inference

Reviewer: ```suggestion
As an example, an end-to-end ResNet-based image classifier contains 3 major parts in its pipeline:
1. A data loader that reads the input images from disk, memory or network
2. A sequence of image transformation that normalizes the input images, including resize, crop, type conversions, etc
3. Finally, a ResNet that maps a batch of input images to their class labels accordingly
```

Author: the first step reads the images and decode image binary( in `png/jpeg/...`）to tensor

Reviewer: Updated. Let me know if it reads good to you :-)

> 

Reviewer: ```suggestion
Below is a snippet that illustrates the workflow of this pipeline:
```

```
class Predictor(nn.Module):

    def __init__(self, tvm_module=None):
        super().__init__()
        self.resnet18 = resnet18(pretrained=True, progress=False).eval()
        self.transforms = nn.Sequential(
            T.Resize([256, ]),  # We use single int value inside a list due to torchscript type restrictions
            T.CenterCrop(224),
            T.ConvertImageDtype(torch.half),
            T.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
        )

    def forward(self, image_path: List[str]) -> torch.Tensor:
        with torch.no_grad():
            images: List[torch.Tensor] = []
            for path in image_path:
                img = read_image(path)
                images.append(img)

>             x = torch.stack(images).cuda().half()

Reviewer: Shall we conclude these lines to an abstract data load to be consistent with our description?

Author: maybe just keep the `class` simple?

            x = self.transforms(x)

>             print(x.shape)

Reviewer: Shall we remove the print here?

Author: ```suggestion
            print(x.shape)
```
```suggestion
```

Author: yes
```suggestion
```

            y_pred = self.resnet18(x)
            return y_pred.argmax(dim=1)
```

> We choose to accelerate resnet model with PyTorchTVM

Reviewer: ```suggestion
With PyTorchTVM, we are able to compile the ResNet with TVM and embed it back to PyTorch seamlessly with a few lines of code:
```

> ```

Author: ```suggestion
``` python
```

```
from tvm.contrib.pt_op import PyTorchTVMModule, compile

print("compile...")
option = {
    "input_infos": [
        ("x", (1, 3, 224, 224)),
    ],
    "default_dtype": "float16",
    "export_dir": "pytorch_compiled",
    "num_outputs": 1,
    "tuning_n_trials": 0,  # set zero to skip tuning
    "tuning_log_file": "tuning.log",
}
x = torch.randn(1, 3, 224, 224).cuda().half()
resnet_jit = torch.jit.trace(model.resnet18, x)
resnet_tvm = compile(resnet_jit, option)
```

> Then we can use the accelerated tvm module directly in pytorch, and also use `torch.jit.script` together with the other 2 parts.

Reviewer: ```suggestion
The TVM-accelerated `resnet_tvm` module can be used directly in PyTorch, or integrated into TorchScript with `torch.jit.script` along with all other PyTorch-native operations.
```

> ```

Author: ```suggestion
``` python
```

```
resnet_tvm = torch.jit.script(resnet_tvm)
print(resnet_tvm.graph)

class PredictorTVM(nn.Module):

    def __init__(self):
        super().__init__()
        self.resnet18 = resnet_tvm
        self.transforms = nn.Sequential(
            T.Resize([256, ]),  # We use single int value inside a list due to torchscript type restrictions
            T.CenterCrop(224),
            T.ConvertImageDtype(torch.half),
            T.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
        )

    def forward(self, image_path: List[str]) -> torch.Tensor:
        with torch.no_grad():
            images: List[torch.Tensor] = []
            for path in image_path:
                img = read_image(path)
                images.append(img)
            x = torch.stack(images).cuda().half()
            x = self.transforms(x)

>             # y_pred = self.resnet18(x)
>             y_pred = self.resnet18([x])[0]

Reviewer: To make it more natural

```suggestion
            y_pred, = self.resnet18([x])
```

Author: Actually, It's a limitation of our custom op interface. So maybe we should keep the diff more obvious to users.

            return y_pred.argmax(dim=1)

print("run tvm...")
model_tvm = PredictorTVM().cuda().half()
for i in range(20):
    t = time.time()
    model_tvm([image_path])
    torch.cuda.synchronize()
    print(time.time() - t)

torch.jit.script(model_tvm).save("model_tvm.pt")
```

> Finally, we get a TVM accelerated model, which can be loaded and served in production.

Reviewer: ```suggestion
Note that the script above provides a seamless serializable solution that allows TVM acceleration to be embedded into TorchScript and thus served in online production without extra effort.
```

> ```

Author: ```suggestion
``` python
```

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

We have opened an initial PR: https://github.com/apache/tvm/pull/8777

The essential cpp code is as follows:

> ```

Author: ```suggestion
``` c++
```

    torch::jit::class_<TvmGraphRuntimeClass>("tvm_class", "TvmGraphModule")
        .def(torch::init<const int64_t, const int64_t, const std::string&>())
        .def("load_tvm_module", &TvmGraphRuntimeClass::LoadTvmModule)
        .def("forward", &TvmGraphRuntimeClass::forward)
        .def("to", &TvmGraphRuntimeClass::to)
        .def_pickle(
            ...
            });
```

And we wrap the custom class in Python:

```
class GraphModule(torch.nn.Module):
    def __init__(self, num_inputs, num_outputs, device=None):
        ...
        self.engine = torch.classes.tvm_class.TvmGraphModule(num_inputs, num_outputs, self.device)
        
    def init(self, input_shapes, lib_path, graph_path, params_path):
        self.engine.load_tvm_module(input_shapes, lib_path, graph_path, params_path)

    def forward(self, inputs: List[torch.Tensor]):
        return self.engine.forward(inputs)
        
    ...
```

# Drawbacks
[drawbacks]: #drawbacks

There are some limitations now:

1. Dynamic shape support

    Currelty we support multiple input_shapes with a bucket policy, which is hacky. A more formal implementation will be in our future work.

2. Zero overhead output

>     Now we only have `set_input_zero_copy`, but our `set_output` has a memcpy.

Author: Yes, it's definitively useful here. I'll update the implementation. Thanks for your information!

> 3. Default performance of TVM

Reviewer: ```suggestion
3. Performance of TVM without tuning
```

    Without autotuning, the performance of TVM is most likely worse compared with native pytorch. To give users immediate feedback, maybe we can make tvm use cudnn/cublas/cutlass as a default implementation.

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

PyTorch is rapidly being adopted because of its user-fiendly API and dynamic capability. With `PytorchTVM`, users can accelerate PyTorch model inference without compromising the full functionality of PyTorch.

# Prior art
[prior-art]: #prior-art

> Our implementation is inspired by this RFC: https://discuss.tvm.apache.org/t/rfc-add-tensorflow-custom-op-to-embed-tvm-runtime-in-tensorflow-graph-and-session/4601

Reviewer: ```suggestion
Our implementation is inspired by this RFC that embeds TVM into TensorFlow: https://discuss.tvm.apache.org/t/rfc-add-tensorflow-custom-op-to-embed-tvm-runtime-in-tensorflow-graph-and-session/4601
```

Author: To clarify: a major difference is that, the TF RFC only support embedding a TVM op, but not a TVM graph. And we also have an internal implementation that embeds TVM graph into TF.

Reviewer: Updated with clarification

> ```

Author: ```suggestion
``` python
```

# Unresolved questions
[unresolved-questions]: #unresolved-questions

> TODO

Reviewer: ```suggestion
See Section Drawbacks
```

>     Currelty we support multiple input_shapes with a bucket policy, which is hacky. A more formal implementation will be in our future work.

Reviewer 2: nit: "currently"

Author: ```suggestion
    Currently we support multiple input_shapes with a bucket policy, which is hacky. A more formal implementation will be in our future work.
```
